examples/loaders/preprocessor-typescript.mjs

```diff
@@ -0,0 +1,71 @@
+import ts from 'typescript';
+
+export let sourceExtensionTypes = ['.ts', '.tsx'];
+
+export let outputExtensionTypes = ['.mjs', '.cjs'];
+
+export function getPreProcessor(configOptions = {}) {
+  return {
+    process(source, url) {
+      // Return new, resolved specifier using ts-specific
+      // logic if one exists; undefined otherwise
+      function getNewSpecficier(oldSpecifier) {
+        let moduleResolutionHost = {
+          fileExists: fileName => {
+            return ts.sys.fileExists(fileName);
+          },
+        };
+        return ts.resolveModuleName(
+          oldSpecifier,
+          url.replace('file://', ''),
+          configOptions.compilerOptions || {},
+          moduleResolutionHost
+        ).resolvedModule?.resolvedFileName;
+      }
+
+      // Transformer to replace import specifiers with actual path
+      // using ts-specific logic
+      function tsModuleResolver() {
+        return context => {
+          function visit(node, inImportExpression) {
+            if (inImportExpression && ts.isStringLiteral(node)) {
+              let oldSpecifier = node.text;
+              let newSpecifier = getNewSpecficier(oldSpecifier);
+              if (newSpecifier) {
+                return ts.createStringLiteral(newSpecifier);
+              }
+              return node;
+            } else if (ts.isImportDeclaration(node) || ts.isImportCall(node)) {
+              return ts.visitEachChild(
+                node,
+                child => visit(child, true),
+                context
+              );
+            } else {
+              return ts.visitEachChild(
+                node,
+                child => visit(child, false),
+                context
+              );
+            }
+          }
+
+          return node => ts.visitNode(node, visit);
+        };
+      }
+
+      // Add resolver to hook into transpiler logic "before"
+      // transpiling occurs
+      configOptions.transformers = {before: [tsModuleResolver()]};
```

Comment:
Style nit: I generally prefer copy over mutate. Especially for arguments it can be problematic to mutate the value - the caller of this function may not expect the object to change.

Alternative:

```js
configOptions = {
  ...configOptions,
  transformers: {before: [tsModuleResolver()]},
};
```

Reply:
Good point, updated this and accounted for caller having their own transformers in `before` as well